File: Python/Pipeline.py
#!/usr/bin/env python3

import os
import sys
import subprocess
from pathlib import Path
from yaml import safe_load


###################################################
##      load packages
###################################################

# Assuming RSV_functions is in the same directory as this script
from RSV_functions import get_sub_folders, elements_not_in_array, check_tool_availability


###################################################
##      load parameters from YAML file
###################################################
config_file = sys.argv[1]
try:
    with open(config_file, 'r') as f :
        config = safe_load(f)
except:
    print('Can not open this file!')
    sys.exit()

data_folder_name = config['DATA_DIR']
reference_folder_name = config['REFERENCE_DIR']
working_folder_name = config['OUTPUT_DIR']

star_ThreadN = config.get('STAR_THREAD_N', 16)
ref_use = config.get('REF_USE', 'all')


###################################################
##      Check tools' availability
###################################################
print(f"######################\tStep 0 Checking tools' availability ... \t######################")
check_tool_availability_res = 0
# Check for IGVtools
check_tool_availability_res += check_tool_availability("igvtools")
# Check for Samtools
check_tool_availability_res += check_tool_availability("samtools")
# Check for fastp
check_tool_availability_res += check_tool_availability("fastp")
# Check for STAR
check_tool_availability_res += check_tool_availability("STAR")

if check_tool_availability_res > 0:
    sys.exit()

print(f"######################\tStep 0  PASSED ... \t######################")


###################################################
##      input data read
###################################################
data_folder_name = os.path.normpath(data_folder_name)

sample_dict = {}

with os.scandir(data_folder_name) as entries:
    for entry in entries:
        if entry.name.endswith('.fastq.gz') and entry.is_file():
            file = entry.name
            sample_id = file.split('_R')[0]

            if sample_id not in sample_dict:
                sample_dict[sample_id] = ['R1','R2']
                
            if '_R1' in file:
                sample_dict[sample_id][0] = file
            elif '_R2' in file:
                sample_dict[sample_id][1] = file


###################################################
##      setup working dir
###################################################
working_folder_name = os.path.normpath(working_folder_name)
if not os.path.exists(working_folder_name):
    os.mkdir(working_folder_name)


###################################################
##      start process
###################################################

for sample_id in sorted(sample_dict.keys()):
    print(f"Start process {sample_id} ...")

    # skip Undetermined reads
    if "Undetermined" in sample_id:
        continue

    cur_folder = os.path.join(working_folder_name, sample_id)
    if not os.path.exists(cur_folder):
        os.mkdir(cur_folder)

    original_read1 = os.path.join(data_folder_name, sample_dict[sample_id][0])
    original_read2 = os.path.join(data_folder_name, sample_dict[sample_id][1])

    
    ##############################################################################
    ### step 1    QC using fastp
    ##############################################################################
    print(f"######################\tStep 1 Quality check and trimming using fastp ... \t######################")

    read1_trim = "reads_trimed_R1.fastq"
    read2_trim = "reads_trimed_R2.fastq"

    cmd = f"fastp -i {original_read1} -I {original_read2} -o {read1_trim} -O {read2_trim}"
    subprocess.run(cmd, shell=True, cwd=cur_folder)

    print("######################\nFinish Quality trimming . \n######################")

    
    ##############################################################################
    ## step 2    STAR mapping
    ##############################################################################
    print(f"######################\tStep 2 STAR mapping ... \t######################")

    # get all reference DB
    all_ref_db = get_sub_folders(reference_folder_name)

    if ref_use != "all":
        ref_use_list = ref_use.split(',')
        not_exist = elements_not_in_array(ref_use_list, all_ref_db)
        if not_exist:
            string = ', '.join(not_exist)
            raise ValueError(f"These references can not be found under your reference folder: {string}")
        else:
            all_ref_db = ref_use_list

    for ref_db in all_ref_db:
        refdb_A = os.path.join(reference_folder_name, ref_db)
        map_A_folder = os.path.join(cur_folder, ref_db)

        if not os.path.exists(map_A_folder):
            os.mkdir(map_A_folder)

        cmd = f"STAR --runThreadN {star_ThreadN} --genomeDir {refdb_A} --outSAMtype BAM SortedByCoordinate --readFilesIn {read1_trim} {read2_trim} --outFileNamePrefix {map_A_folder}/"
        subprocess.run(cmd, shell=True, cwd=cur_folder)

    print("######################\nFinish STAR mapping . \n######################")

    
    ##############################################################################
    ## step 3    index and Count using Samtools and IGVtools 
    ##############################################################################
    print(f"######################\tStep 3 Index and Count ... \t######################")

    for ref_db in all_ref_db:
        map_A_folder = os.path.join(cur_folder, ref_db)
        ref_db_folder = os.path.join(reference_folder_name, ref_db)

        cmd_index = f"samtools index Aligned.sortedByCoord.out.bam"
        subprocess.run(cmd_index, shell=True, cwd=map_A_folder)

        cmd_count = f"igvtools count -z 5 -w 1 --bases Aligned.sortedByCoord.out.bam alignments.cov.wig {ref_db_folder}.fasta"
        subprocess.run(cmd_count, shell=True, cwd=map_A_folder)

    print("######################\nFinish Index and Count \n######################")


###################################################
##      program finished
###################################################

print("MApping finished!")

File: Perl/Pipeline.pl
#!/usr/bin/perl

BEGIN {unshift @INC,"/research/groups/cab/projects/automapper/common/lli75/RSV"};

###################################################
##      load packages
###################################################

use strict;
use YAML::XS 'LoadFile';

# load RSV functions
use RSV_functions;


###################################################
##      load parameters from YAML file
###################################################
my $config_file = shift;
my $config = LoadFile($config_file) or die "Can not open $config_file as a YAML file!";

my $data_folder_name = $config->{'DATA_DIR'} or die "Please set up a DATA_DIR!";
my $reference_folder_name = $config->{'REFERENCE_DIR'} or die "Please set up a REFERENCE_DIR!";
my $working_folder_name = $config->{'OUTPUT_DIR'} or die "Please set up a OUTPUT_DIR!";

my $star_ThreadN = $config->{'STAR_THREAD_N'} // 16;
my $ref_use = $config->{'REF_USE'} // 'all';


###################################################
##      input data read
###################################################
$data_folder_name =~ s/\/\//\//g;

my %sample_hash;

opendir(my $dh, $data_folder_name) or die "Can not open $data_folder_name: $!";
while (readdir $dh) 
{
	if($_ =~ /\.fastq/)
	{
		my $file = $_;
		my $id = $file;
		$id =~ s/_R\d.+//;
		if($file =~ /_R1/)
		{
			$sample_hash{$id}{1} = $file;
		}
		elsif($file =~ /_R2/)
		{
			$sample_hash{$id}{2} = $file;
		}
	}
}
closedir $dh;


###################################################
##      setup working dir
###################################################
$working_folder_name =~ s/\/\//\//g;
if(!-e $working_folder_name)
{
	mkdir($working_folder_name) or die "Can not create the $working_folder_name!";
}


###################################################
##      start process
###################################################

foreach (sort {$a <=> $b} keys %sample_hash)
{
	print STDERR "Start process $_ ... \n";

	# skip Undetermined reads
	if ($_ =~ /Undetermined/) 
	{
		next;
	}

	my $cur_folder = $working_folder_name."/".$_;
	if(!-e $cur_folder)
	{
		mkdir($cur_folder) or die "Can not create the $cur_folder! Check the path and premission!";
	}

	my $original_read1 = $data_folder_name."/".$sample_hash{$_}{1};
	my $original_read2 = $data_folder_name."/".$sample_hash{$_}{2};

	
	##############################################################################
	### step 1    QC using fastp
	##############################################################################
	print STDERR "######################\tStep 1 Quality check and trimming using fastp ... \t######################\n";

	my $read1_trim = "reads_trimed_R1.fastq";
	my $read2_trim = "reads_trimed_R2.fastq";

	my $cmd = "cd $cur_folder; fastp -i $original_read1 -I $original_read2 -o $read1_trim -O $read2_trim";
	# print STDERR $cmd;
	`$cmd`;

	print STDERR "######################\nFinish Quality trimming . \n######################\n";

	
	##############################################################################
	## step 2    STAR mapping
	##############################################################################
	print STDERR "######################\tStep 2 STAR mapping ... \t######################\n";

	# get all reference DB
	my @all_ref_db = get_sub_folders($reference_folder_name);

	if ($ref_use ne "all") {
		my @ref_use_list = split(',',$ref_use);
		my @not_exist = elements_not_in_array(\@ref_use_list, \@all_ref_db);
		if (@not_exist) {
			my $string = join ',', @not_exist;
			die "These references can not be found under your reference folder: $string\n";
		} else {
			@all_ref_db = @ref_use_list;
		}
	}

	foreach my $ref_db (@all_ref_db) {
		my $refdb_A = "$reference_folder_name/$ref_db";

		my $map_A_folder = $cur_folder.'/$ref_db/';

		if(!-e $map_A_folder)
		{
			mkdir($map_A_folder) or die "Can not create the $map_A_folder! Check the path and premission!";
		}

		my $cmd = "cd $cur_folder; STAR --runThreadN $star_ThreadN --genomeDir $refdb_A --outSAMtype BAM SortedByCoordinate --readFilesIn $read1_trim $read2_trim --outFileNamePrefix $map_A_folder";
		# print STDERR $cmd;
		`$cmd`;
	}

	print STDERR "######################\nFinish STAR mapping . \n######################\n";

	
	##############################################################################
	## step 3    index and Count using Samtools and IGVtools 
	##############################################################################
	print STDERR "######################\tStep 3 Index and Count ... \t######################\n";

	
	foreach my $ref_db (@all_ref_db) {
		my $map_A_folder = $cur_folder.'/$ref_db/';
		my $refdb_A = "$reference_folder_name/$ref_db";

		my $cmd = "cd $map_A_folder; samtools index Aligned.sortedByCoord.out.bam";
		# print STDERR $cmd;
		`$cmd`;

		my $cmd = "cd $map_A_folder; igvtools count -z 5 -w 1 --bases  Aligned.sortedByCoord.out.bam  alignments.cov.wig  $refdb_A.fasta";
		# print STDERR $cmd;
		`$cmd`;
	}

	print STDERR "######################\nFinish Index and Count \n######################\n";

}

